### PR TITLE
Add ability to ignore failures on certain environments

### PR DESCRIPTION
**Pull request by MarkRx**
Adds ability to configure a regular expression to indicate environments that can ignore errors. If an environment ignores errors the UI will use a gray icon instead of a red one. The deploy widget will only turn red if an environment that is not ignoring errors fails.

Use case: There are cases when we don't care if an environment is down. In a blue/green stage-active deployment we never care about the "staging" environment. Without this change the deploy widget will almost always be red (as usually no code is staged).

** Update **
Configuration is specified in the deploy widget config

Requires PR #891 

taxsdlc666-1
taxsdlc-666-2

Update: Added additional ability to configure on a per dashboard basis.
taxsdlc666-4

### CODE REVIEW

**Comment by tabladrum** [comment]
@MarkRx we might have merge some PR's out of order. Can you see if this conflict can be resolved?

**Comment by tabladrum** [comment]
@MarkRx this needs few quick manual merging.

**Comment by MarkRx** [comment]
rebased & squashed

**Comment by tabladrum** [comment]
Looks like you are saving this config as a system global config. Better if we save it within the widget config. And then modify the deploy widget to have the user to enter the regex pattern. We don't want the same regex affecting every user or we don't want to change the properties file everytime there is a new regex.

Thoughts?

**Comment by MarkRx** [comment]
Will look at next week.

**Comment by MarkRx** [comment]
@tabladrum done. Screenshot above.

Rebased.

**Comment by MarkRx** [comment]
@pxk5958

**Comment by MarkRx** [comment]
Done. Will go through an rebase everything once #1042 is merged in.

**Comment by MarkRx** [comment]
rebased onto master 2.0.5

File: UI/src/app/dashboard/directives/dash-status.js [hunk 1]
```diff
@@ -55,6 +58,11 @@
                     case 'warning':
                         scope.currentStatus = DashStatus.WARN;
                         break;
+                    case 0:
+                    case '0':
```
Comment (tabladrum) [inline UI/src/app/dashboard/directives/dash-status.js]: what are these 0 and '0' status?

File: UI/src/components/widgets/deploy/module.js [hunk 2]
```diff
@@ -21,6 +21,29 @@
 
     angular
         .module(HygieiaConfig.module)
+        .directive('validregex', function() {
+          return {
+            require: 'ngModel',
+            link: function(scope, elm, attrs, ctrl) {
+              ctrl.$validators.validregex = function(modelValue, viewValue) {
+                if (ctrl.$isEmpty(modelValue)) {
```
Comment (tabladrum) [inline UI/src/components/widgets/deploy/module.js]: Do we need this check? I think the RegExp(..) will handle the empty scenario too?

**Comment by nireeshT** [comment]
@MarkRx @tabladrum  - Does the regex to ignore failures check for environment names and deployment job names.

**Comment by tabladrum** [comment]
@nireeshT just the environment names.

File: UI/src/app/dashboard/directives/dash-status.js (hunk 1 above)
Comment (nireeshT) [inline UI/src/app/dashboard/directives/dash-status.js]: @tabladrum  - 0 or '0' status is newly added IGNORE: 0

Comment (tabladrum) [inline UI/src/app/dashboard/directives/dash-status.js]: Now I see it. Sorry.

**Pull request merged**